oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -117,7 +90,101 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
      * Validates access token and refreshes it if necessary.
      */
     @Override
-    public final CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
-        return tokenManager.get();
+    public final CompletionStage<GrantedOAuth2AccessToken> getAccessToken(ClientRequestContext ctx) {
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
+        doGetAccessToken(ctx, future, 1);
+        return future;
+    }
+
+    private void doGetAccessToken(ClientRequestContext ctx,
+                                  CompletableFuture<GrantedOAuth2AccessToken> future, int attempts) {
+        final GrantedOAuth2AccessToken accessToken = this.accessToken;
+
+        if (isValidToken(accessToken)) {
+            future.complete(accessToken);
+            return;
+        }
+        if (!authenticatingUpdater.compareAndSet(this, 0, 1)) {
+            scheduleNextRetry(ctx, future, attempts);
```

Comment:
If this condition is false, it means that another executor is authenticating to get the access token. So I think we don't have to schedule the next retry but just use the token when it's ready.
How about changing:
```java

private static final AtomicReferenceFieldUpdater<AbstractOAuth2AuthorizationGrant, CompletableFuture> accessTokenFutureUpdater = ....

private volatile CompletableFuture<GrantedOAuth2AccessToken> accessTokenFuture;

private void doGetAccessToken(...) {

    ...    
    if (!accessTokenFutureUpdater.compareAndSet(this, null, future)) {            
        accessTokenFuture.handle((...) -> {... future.complete(token)...});
    }
    ...
}
```